src/main/scala/akka/stream/contrib/Retry.scala

```diff
@@ -0,0 +1,244 @@
+    GraphDSL.create(flow) { implicit b => origFlow =>
+      import GraphDSL.Implicits._
+
+      val retry = b.add(new RetryCoordinator[I, S, O](retryWith))
+
+      retry.out2 ~> origFlow ~> retry.in2
+
+      FlowShape(retry.in1, retry.out1)
+    }
+  }
+
+  /**
+   * Factory for multiple retries flow. similar to the simple retry, but this will allow to
+   * break down a "heavy" element which failed into multiple "thin" elements, that may succeed
+   * indevidually. Since it's very easy to inflate elements in retry cycle, there's also a
+   * limit parameter, that will limit the amount of generated elements by the `retryWith` function,
+   * and will fail the stage if that limit is exceeded.
+   * Passing `Some(Nil)` is valid, and will result in filtering out the failure quietly, without
+   * emitting a failed `Try` element.
+   *
+   * IMPORTANT CAVEAT:
+   * The given flow must not change the number of elements passing through it (i.e. it should output
+   * exactly one element for every recieved element). Ignoring this, will have an unpredicted result.
+   *
+   * @param limit since every retry can generate more elements,
+   *        the inner queue can get too big. if the limit is reached,
+   *        the stage will fail.
+   * @param flow the flow to retry
+   * @param retryWith if output was failure, we can optionaly recover from it,
+   *        and retry with a sequence of input & new state pairs we get from this function.
+   * @tparam I input elements type
+   * @tparam O output elements type
+   * @tparam S state to create a new (input,state): (I,S) to retry with
+   * @tparam M materialized value type
+   * @return `Graph[FlowShape[(I,S),Try[O]],M]` instance
+   */
+  def concat[I, O, S, M](limit: Long, flow: Graph[FlowShape[(I, S), (Try[O], S)], M])(retryWith: S => Option[immutable.Iterable[(I, S)]]): Graph[FlowShape[(I, S), (Try[O], S)], M] = {
+    GraphDSL.create(flow) { implicit b => origFlow =>
+      import GraphDSL.Implicits._
+
+      val retry = b.add(new RetryConcatCoordinator[I, S, O](limit, retryWith))
+
+      retry.out2 ~> origFlow ~> retry.in2
+
+      FlowShape(retry.in1, retry.out1)
+    }
+  }
+
+  private[akka] class RetryCoordinator[I, S, O](retryWith: S => Option[(I, S)]) extends GraphStage[BidiShape[(I, S), (Try[O], S), (Try[O], S), (I, S)]] {
+    val in1 = Inlet[(I, S)]("RertryCoordinator.in")
+    val out1 = Outlet[(Try[O], S)]("RertryCoordinator.out")
+    val in2 = Inlet[(Try[O], S)]("RertryCoordinator.cin")
+    val out2 = Outlet[(I, S)]("RertryCoordinator.cout")
+    override val shape = BidiShape[(I, S), (Try[O], S), (Try[O], S), (I, S)](in1, out1, in2, out2)
+    override def createLogic(attributes: Attributes) = new GraphStageLogic(shape) {
+      var elementInCycle = false
+      var pending: (I, S) = null
+
+      setHandler(in1, new InHandler {
+        override def onPush() = {
+          val is = grab(in1)
+          if (!hasBeenPulled(in2)) pull(in2)
```

Comment:
hm, does not make sense instead to pull through the wrapped stage? I.e. pulling of out1 results this stage pulling the wrapped flow, which in turn will pull out2, which in turn will pull in1? That is a control flow that might be more understandable.